draft-ietf-quic-multipath.md

```diff
@@ -377,10 +377,15 @@ When the multipath extension is negotiated, a client that wants to use an
 additional path MUST first initiate the Address Validation procedure
 with PATH_CHALLENGE and PATH_RESPONSE frames as described in
 {{Section 8.2 of QUIC-TRANSPORT}}, unless it has previously validated
-that address. After receiving packets from the
+that address.
+
+After receiving packets from the
 client on a new path, if the server decides to use the new path,
 the server MUST perform path validation ({{Section 8.2 of QUIC-TRANSPORT}})
 unless it has previously validated that address.
+An endpoint that receives a PATH_CHALLENGE and does not want to establish
+this path should immediately close the path by sending a PATH_ABANDON
```

Comment:
Shouldn't be this "should" normative?

Reply:
@huitema and I are still discussing this. Please see discussion in the conversations. I will relabel as not not-ready-to-merge.

Reply:
@mirjak the conversations are marked as "resolved", so it is hard for @qdeconinck and others to get the arguments. (This is a side effect of carrying discussions in PR instead of in a specialized issue.)

@qdeconinck I think the text is fine, but we have two items of disagreement.

1. Should we use normative language (e.g., should -> SHOULD) in this specific section? 

2. If we do use normative language, should we say SHOULD or MUST?

The argument against normative language that we are going to make a reference to {{close}}, which itself contains normative language, so we should not duplicate it. I do not think this is a strong argument, because the paragraph just before this says _"When the multipath extension is negotiated, a client that wants to use an
additional path MUST first initiate the Address Validation procedure
with PATH_CHALLENGE and PATH_RESPONSE frames as described in
{{Section 8.2 of QUIC-TRANSPORT}}..."_ We also know from experience that using lower case "should" tends to confuse the readers, who will wonder whether this is a hard requirement or not.

The argument for MUST is the classic issue: if you say "SHOULD" instead of "MUST", in which cases shall the implementation not follow your advice? If you cannot spell out a case, just say MUST. In this case, there are plausible arguments for waiting a bit in some cases, e.g., waiting until an appropriate CID becomes available. That CID may well be just following the PATH_CHALLENGE in the same packet, or it may be arriving on another path. My original proposal had text to that effect, but we split it out.

So, my preference would be for SHOULD (upper case), and some additional text explaining the alternative. I think the additional text should be as "matter of fact" as possible. Something like: "The endpoint MAY delay sending a PATH_RESPONSE or a PATH_ABANDON, knowing that if neither or those is sent in a reasonable time the peer will eventually close the path as specified in section {{close}}."

Reply:
Just on your first point: I don't understand your argument. This section has normative language; that's fine because it's the normative language on what you need to do when you open a path. However, the normative language on closing a path should be in the section on path closure. Also what I wanted to say is that the normative requirement is already there in that section on path closure and broad enough that is already covers this case ("An endpoint that wants to close a path MUST explicitly terminate the path by sending a PATH_ABANDON frame."). It's not only about saying the same things twice normatively in the document in different places (which I think is also bad) but also to make clear that this case here is actually not special. It's always the same: if you want to close a path you have to send a PATH_ABANDON no matter what.

However, I think the point you are trying to make about the sentence here is actually slightly different because it says "should immediately close". So the question with the word "immediately" in there is: do we want to give any (normative) recommendation when exactly to close the path? First, we can't really enforce that an endpoint sends anything immediately or not anyway. However, I also don't understand your point above why there should be any benefit in waiting...? If you don't send the PATH_RESPONSE immediately because you are hoping to get more ressources, then you actually don't want to close the path (and respectively should not send an PATH_ABANDON). However, if you have decided that you don't want to establish the path (and that's what the sentence says), I don't see any reason to delay the sending of a PATH_ABANDON.

In summary I don't see a reason to delay the PATH_ABANDON and I don't think we should talk about the timing at all. So I still don't think we should use normative language here. However, I agree that lower case "should"s are confusing and we could consider using a different phrasing that avoids that!

Reply:
Maybe the simplest solution is to strike "immediately" and just say SHOULD. And yes, you are correct about the gray area between "decide it does not want a path" and "decide it wants it but cannot accept just yet" and "accept the path". Maybe that's something best explained in the implementation section, not here.